### PR TITLE
EDSC-3635: As a cmr-graphql user, I want to retrieve Order Options

### PR DESCRIPTION
**Pull request by eudoroolivares2016**
# Overview

### What is the feature?

Adding the order-options concept to graphql these are for the purpose of retiring legacy services by exposing these concepts to graphql to be called from EDSC

### What is the Solution?

Added the concept and its accompanying fields. order-options can be retrieved from services in cases such as 
`query ($params: CollectionsInput) {
collections(params: $params) {
  items {
    services {
      items {
        type
        orderOptions {
          count
          items {
            conceptId
            form
          }
        }
      }
    }
  }
}}`
possible given a concept-id or some identifying collection parameter

### What areas of the application does this impact?

new concept is decoupled from the 

# Testing

### Reproduction steps

local graphdb instance testing against the cmr SIT env find or ingest a service, collection, and order-option
associate the service to the collection and associate the order-option to the service. Query graphql with the query above

Will add further details in the test session

### Attachments

No relevant screenshots

# Checklist

- [X] I have added automated tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings

### CODE REVIEW

File: src/cmr/concepts/collection.js [hunk 1]
```diff
@@ -21,6 +21,7 @@ export default class Collection extends Concept {
    * @param {String} id Concept ID to set a value for within the result set
    * @param {Object} item The item returned from the CMR json endpoint
    */
+  // TODO: Since all concepts now support associations we should refactor this on the parent class
```
Comment (jaybarra) [inline src/cmr/concepts/collection.js:24]: What would be the LOE to take care of that in this ticket so we can take care of the TODO?

File: src/cmr/concepts/orderOption.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  /**
+   * Instantiates an OrderOption object
+   * @param {Object} headers HTTP headers provided by the query
+   * @param {Object} requestInfo Parsed data pertaining to the Graph query
+   * @param {Object} params GraphQL query parameters
+   */
+  constructor(headers, requestInfo, params) {
+    // This concept uses the "-" character to delineate spaces in CMR we must pass it
+    // in this form to fetch order option concepts from CMR
+    super('order-options', headers, requestInfo, params)
```
Comment (jaybarra) [inline src/cmr/concepts/orderOption.js:13]: Should order-options be singular?

File: src/cmr/concepts/orderOption.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  }
+
+  /**
+   * Parse and return the array of data from the nested response body
+   * @param {Object} jsonResponse HTTP response from the CMR endpoint
+   */
+  parseJsonBody(jsonResponse) {
+    const { data } = jsonResponse
+
+    const { items } = data
```
Comment (jaybarra) [inline src/cmr/concepts/orderOption.js:23]: We should have some null checks for safety.

**Review by jaybarra — APPROVED** [review]

File: src/types/orderOption.graphql [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+type Query {
+  orderOptions(
+    "order option's query parameters"
+    params: OrderOptionsInput
+  ): OrderOptionList!
+
+  orderOption(
+    "order option's query parameters"
+    params: OrderOptionInput
+    "The unique concept id assigned to the order-option"
```
Comment (eudoroolivares2016) [inline src/types/orderOption.graphql:54]: remove this it is going to be deprecated

File: src/cmr/concepts/orderOption.js (hunk 2 above)
Comment (eudoroolivares2016) [reply src/cmr/concepts/orderOption.js:13]: In this case it has to be plural because it's going off to` cmr/search/order-options`

File: src/cmr/concepts/collection.js (hunk 1 above)
Comment (eudoroolivares2016) [reply src/cmr/concepts/collection.js:24]: Changing this would only allow us to set associations as a default component of all the concepts, we would still have to explicitly write code to allow the associations to flow through. That would be a different ticket which would be to refactor graphql so that you can add an association just by adding the type to the .graphql file thus reducing some boilerplate.

File: src/cmr/concepts/orderOption.js (hunk 3 above)
Comment (eudoroolivares2016) [reply src/cmr/concepts/orderOption.js:23]: Yea I'm taking a closer look at this one. This is overriding the parent classes method so `orderOption.js `(the concept definition this file is in) doesn't have an accompanying test file since it is largely just overriding the concept class the same is true of all concepts in graphql, I imagine that is why none of these have a default null check case since that would require making a new test file for each of the concepts to get test coverage on this possible branch . The `parseUmmBody` method in the concept class is equivalent to this overridden `parseJsonBody` to handle cmr responses. This method only gets called once in the `parseJson` method from the concept class which is in turn called in another method which has a guard around the null case and passes the response down to here
```
 if (jsonResponse) {
        await this.parseJson(jsonResponse, jsonKeys)
      }
```
which is being covered by tests. Do you think I should add tests for all the concepts for the null case?

File: src/resolvers/__tests__/grid.test.js [hunk 5]
```diff
@@ -185,7 +185,7 @@ describe('Grid', () => {
                 name: 'Lorem Ipsum'
               }]
             })
-
+            // TODO remove this parameter search from grids in favor of params
```
Comment (eudoroolivares2016) [inline src/resolvers/__tests__/grid.test.js:188]: this is for grids which are going to need to be updated in the future since ingest team is currently updating their schema

File: src/types/orderOption.graphql [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  "The description for this order option"
+  description: String
+  "The ECHO Forms schema for this order option"
+  form: String
+  "The guid-id for this order option"
+  id: String!
+  "The orderOption name."
+  name: String
+  "The native id to set on the order option."  
+  nativeId: String
```
[the remark this answers is not in the document]
Comment (eudoroolivares2016) [reply src/types/orderOption.graphql:15]: I am going to leave it in for now. I think this is an excellent point but, I do think we ought to keep things somewhat consistent. Since the `native-id` field is returned in all the other concepts and, at least as I understood from some discussion last week, one client uses `native-id `in their search (cmr-stac) for some concepts. We should consider whether to remove this field more broadly though

File: src/resolvers/__tests__/grid.test.js (hunk 5 above)
Comment (macrouch) [reply src/resolvers/__tests__/grid.test.js:188]: I don't understand this TODO, is it just switching from the deprecated option to the params object? Or is something changing soon with grids? 

If something is changing we should get a ticket number to add to the comment.

File: src/cmr/concepts/orderOption.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+  /**
+   * Query the CMR UMM API endpoint to retrieve requested data
+   * @param {Object} searchParams Parameters provided by the query
+   * @param {Array} requestedKeys Keys requested by the query
+   * @param {Object} providedHeaders Headers requested by the query
+   */
+  fetchUmm(searchParams, ummKeys, headers) {
+    const ummHeaders = {
+      ...headers,
+      Accept: 'application/vnd.nasa.cmr.umm_results+json'
```
[the remark this answers is not in the document]
Comment (eudoroolivares2016) [reply src/cmr/concepts/orderOption.js:37]: Yea about that... So generics don't support versioning currently. That is a ticket I imagine that ingest has now inherited from UMM_powerhouse. Perhaps I should leave a TODO here so that when CMR supports that we can make sure we stay updated with the latest version?

File: src/resolvers/service.js [hunk 8]
```diff
@@ -35,6 +35,52 @@ export default {
...
+      const { collections = [] } = associationDetails
+      // If there are no associations to collections for this service
+      if (!collections.length) {
+        return {
+          count: 0,
+          items: null
+        }
+      }
+
+      // eslint-disable-next-line max-len
```
[the remark this answers is not in the document]
Comment (eudoroolivares2016) [reply src/resolvers/service.js:66]: 🤦

File: src/resolvers/__tests__/grid.test.js (hunk 5 above)
Comment (eudoroolivares2016) [reply src/resolvers/__tests__/grid.test.js:188]: No this is just referring to using params object instead of the deprecated obj in the tests. I should write that ticket to update grid though now that the changes are in CMR and that's where I would make those test modifications.

File: src/cmr/concepts/collection.js (hunk 1 above)
Comment (eudoroolivares2016) [reply src/cmr/concepts/collection.js:24]: Looking at this I will go ahead and write a ticket for it. I think we'll want to add `association_details` to all the concepts as well since they all support that field. EXCEPT granules. I am a bit concerned looking at it closer that it will cause an issue on granule searching with `graphql` I don't THINK it will, since the field will just always be null the and get caught, the same way it works for concept's which support `association_details` but, are not associated to anything and thus, don't actually have the field in the response.

File: src/resolvers/__tests__/grid.test.js (hunk 5 above)
Comment (macrouch) [reply src/resolvers/__tests__/grid.test.js:188]: I think you can probably just remove the TODO, deprecated params are used all over the tests. A ticket to update them would be good

File: src/resolvers/service.js [hunk 9]
```diff
@@ -35,6 +35,52 @@ export default {
...
+
+      // Pull out the service's parent collection id to filter the associated orderOptions only to those collections
+
+      const {
+        associationDetails = {},
+        parentCollectionConceptId
+      } = source
+
+      // order-options on the payload are only in the context of the collection to service association
+      if (!parentCollectionConceptId) {
```
[the remark this answers is not in the document]
Comment (eudoroolivares2016) [reply src/resolvers/service.js]: Latest uses this behavior

File: src/resolvers/__tests__/service.test.js [hunk 10]
```diff
@@ -342,5 +344,650 @@ describe('Service', () => {
         }
       })
     })
+
+    // Tests for the associated legacy services order-options
+    test('only retrieve one OO filtered from assoc details', async () => {
```
Comment (macrouch) [inline src/resolvers/__tests__/service.test.js:349]: I think these tests could be organized a little better, add a couple of `describe` blocks. 'When the services are requested with a parent collection query', 'When the services are requested as the top level query'

**Review by macrouch — APPROVED** [review]

**Comment by macrouch** [comment]
Also, please do a 'squash and merge' when you merge, if it lets you

File: src/resolvers/__tests__/service.test.js (hunk 10 above)
Comment (eudoroolivares2016) [reply src/resolvers/__tests__/service.test.js]: Agree thank you!
